tools/ci_build/reduce_op_kernels.py

```diff
@@ -25,17 +27,81 @@
 log = get_logger("reduce_op_kernels")
 
 
+def _adapt_filters_for_extended_minimal_build(base_required_ops: typing.Optional[dict],
+                                              base_op_type_impl_filter: typing.Optional[OpTypeImplFilterInterface]):
+    '''
+    Adapts the values returned by parse_config() for an extended minimal build or higher.
+    In particular:
+    - Includes ONNX ops needed by layout transformation
+    '''
+    # layout transformation requires certain ONNX ops to be available
+    layout_transformation_required_ops = dict()  # op name -> set of opset versions
+    layout_transformation_required_ops_file = ORT_ROOT / "onnxruntime/core/framework/kernel_def_hash_helpers.cc"
+    with open(layout_transformation_required_ops_file, mode="r") as f:
+        region_boundary_pattern = re.compile(r'@@region_(begin|end)\(layout_transformation_required_kernels\)@@')
+        op_to_hash_pattern = re.compile(r'\{"(\w+)_(\d+)",\s+\w+\},')
+        in_region = False
+        for line in f:
+            region_boundary_match = region_boundary_pattern.search(line)
+            if region_boundary_match:
+                in_region = region_boundary_match.group(1) == "begin"
+                continue
+
+            if not in_region:
+                continue
+
+            op_to_hash_match = op_to_hash_pattern.search(line)
+            if op_to_hash_match:
+                op_name, opset = op_to_hash_match.group(1, 2)
+                layout_transformation_required_ops.setdefault(op_name, set()).add(int(opset))
+
+    adapted_required_ops = None
+    if base_required_ops is not None:
+        adapted_required_ops = base_required_ops.copy()
+        required_onnx_ops = adapted_required_ops.setdefault('ai.onnx', dict())
+        for op_type, opsets in layout_transformation_required_ops.items():
+            for opset in opsets:
+                required_onnx_opset_ops = required_onnx_ops.setdefault(opset, set())
+                required_onnx_opset_ops.add(op_type)
+
+    adapted_op_type_impl_filter = None
+    if base_op_type_impl_filter is not None:
+        class _AdaptedFilter(OpTypeImplFilterInterface):
+            def __init__(self, filter_to_adapt: OpTypeImplFilterInterface,
+                         required_optypes: typing.Set[str]):
+                self.filter_to_adapt = filter_to_adapt
+                self.required_optypes = required_optypes
+
+            def is_typed_registration_needed(self, domain: str, optype: str, type_registration_str: str):
+                # Always require registration for ONNX ops in self.required_optypes.
+                if domain == 'ai.onnx' and optype in self.required_optypes:
+                    return True
+                return self.filter_to_adapt.is_typed_registration_needed(domain, optype, type_registration_str)
+
```

Comment:
side note: if needed in the future we could probably check the model for the small set of layout sensitive operators to determine the exact types required. I don't think it's worth doing currently though.